### DISCUSSION TITLE
Help with previousStep and parameters

### DISCUSSION THREAD

**Discussion by SigalZ** [spatie/laravel-livewire-wizard#85] [Q&A]
I have a problem using the previousStep function.

In the "wizard" blade:

```
<x-admin_layout>
    <livewire:product-wizard />
</x-admin_layout>
```

My first step gets a Product model parameter on the mount function, this can be empty when creating a product or a product when updating a product:

```
class ProductDetailsStepComponent extends StepComponent
{
    public $pageTitle="Create a Product";
    public $createProduct = false;
    public Product $product; 
    public $selectedTab;    
    public $alertMsg = 'created';
    public $productTypes;
        
    protected array $rules = [
        'product.product_type_id' => '',
        'product.name' => 'required',
        'product.product_desc' => '',
        'product.out_of_stock' => '',
        'product.new' => '',
        'product.publish' => ''        
    ];
    
    protected array $messages = [
        'product.name.required' => 'Product name cannot be empty',
    ];
  
    public function mount(Product $product)
    {
          $this->product = $product;
	...
}
```

On the blade of the second step I have this button:
```
<button type="button" class="btn btn-light me-4" wire:click="previousStep">Previous Step</button>
```

Pressing this button I get this error:

Cannot assign array to property App\Http\Livewire\Admin\Product\Steps\ProductDetailsStepComponent::$product of type App\Models\Product

I could not find a solution reading the package documents.

I tried this:

In the Wizard component:
```
class ProductWizardComponent extends WizardComponent
{
    public Product $product;

    public function mount(Product $product)
    {
        $this->product = $product;
    }
    
    public function initialState(): array
    {
        return [
            'product-details-step' => [
                'product.product_type_id' => 1,
                'product.name' => '',
                'product.product_desc' => '',
                'product.out_of_stock' => 0,
                'product.new' => 1,
                'product.publish' => 0
            ]
            ];
    }
```

I get the same error

I tried to change the call on the "wizard" blade to:

```
<x-admin_layout>
    <livewire:product-wizard product=0 />
</x-admin_layout>
```
I get this error:

Cannot assign string to property App\Http\Livewire\Admin\Product\ProductWizardComponent::$product of type App\Models\Product

I just don't understand how to do it or if it is possible at all?

**Comment by ju5t** [comment]
If you switch `$product` to an array it will work. You need an array as the wizard has no type support in its hydration process. In Livewire v3 it'll work similarly as binding models will most likely be deprecated.

**Reply by SigalZ** [reply]
Thank you for taking the time to answer me. I just don't understand where should I switch the $product to array? In my: ProductDetailsStepComponent? so I can't use the bind of the model and have to create an array and update the array with the product details when it is an update? How will I send the id of the updated product? I'm sorry, I just need a bit more help, not sure what you mean.

**Reply by ju5t** [reply]
In `ProductDetailsStepComponent` you're passing a `Product`. You could pass a product id instead (and lock it in Livewire 3 so it cannot be changed).

```php
public ?int $productId;

public array $data = [];

public function mount(?int $productId): void
{
    $this->data = $this->product->toArray();
}

public function getProductProperty(): Product
{
    if (empty($this->productId)) {
        return new Product;
    }

    return Product::find($this->productId);
}
```

This stores the state in the public property `$data`. You can wire your fields to that. Your `initialState` should be rewritten accordingly. And when you want to save it, you can do:

```php
$this->product->fill($data)->save();
```

**Reply by SigalZ** [reply]
Thank you very much, I really appreciate your help.
